kubeval/kubeval.go

```diff
@@ -133,18 +144,29 @@ func validateResource(data []byte, schemaCache map[string]*gojsonschema.Schema,
 }
 
 func validateAgainstSchema(body interface{}, resource *ValidationResult, schemaCache map[string]*gojsonschema.Schema, config *Config) ([]gojsonschema.ResultError, error) {
+	var schema *gojsonschema.Schema
+	var err error
+	var ok bool
+
+	cacheMu.Lock()
+	schema, ok = schemaCache[resource.VersionKind()]
+	cacheMu.Unlock()
+	if !ok {
+		schema, err = downloadSchema(resource, config)
```

Comment:
As this isn't guarded by a mutex, this will cause the same schema to be downloaded multiple times. Ideally we would only fetch each schema once as those downloads are pretty slow.

Reply:
I reworked a bit the implementation of the caching, as a precursor to the multi-worker setup... it separates the caching logic from the manifest downloading, and makes it easier to gate by a mutex. https://github.com/instrumenta/kubeval/compare/master...yannh:run-with-workers-rework?expand=1    cc @garethr wdyt? I'd need to adapt the tests.